Tcp egress blog

vadimeisenbergibm opened the pull request:
The compiled version is here https://vadimeisenbergibm.github.io/blog/2018/egress-tcp.html .

geeknoid commented on _blog/2018/egress-tcp.md:
```diff
@@ -0,0 +1,306 @@
+---
```
Please exclude the .DS_Store binary file that slipped in your commit.

vadimeisenbergibm replied:
Done.

geeknoid commented on _blog/2018/egress-tcp.md:
```diff
@@ -0,0 +1,306 @@
+## Bookinfo Sample Application with External Ratings Database
+First, I set up a MySQL database instance to hold book ratings data, outside my Kubernetes cluster. Then I modify the [Bookinfo Sample Application]({{home}}/docs/guides/bookinfo.html) to use my database.
+
+### Setting up the Database for Ratings Data
```
For all the headers in this post, and in your previous post, please use initial caps only, which is our convention:

## Setting up the database for ratings data

vadimeisenbergibm replied:
@geeknoid thank you for the review!
A question - when citing another blog post on istio.io or a task, should I use the capital letters, as in the title of the task, or lower case letters? For example, should it be:

`the  [Consuming External Web Services](https://istio.io/blog/2018/egress-https.html) blog post`
the  [Consuming External Web Services](https://istio.io/blog/2018/egress-https.html) blog post

or 

`the  [consuming external web services](https://istio.io/blog/2018/egress-https.html) blog post`
the  [consuming external web services](https://istio.io/blog/2018/egress-https.html) blog post

geeknoid replied:
You should use the capitalized form that matches the title of the target.

vadimeisenbergibm replied:
Thanks, it clear now.

geeknoid commented on _blog/2018/egress-tcp.md:
```diff
@@ -0,0 +1,306 @@
+---
+{% include home.html %}
+
+In my previous blog post, [Consuming External Web Services]({{home}}/blog/2018/egress-https.html), I described how external services can be consumed by in-mesh Istio applications via HTTPS. In this post, I demonstrate consuming external services over TCP. I use the [Istio Bookinfo Sample Application]({{home}}/docs/guides/bookinfo.html), the version in which the book ratings data is persisted in a ...
```
Here and elsewhere, Sample Application -> sample application

geeknoid commented on _blog/2018/egress-tcp.md:
```diff
@@ -0,0 +1,306 @@
+### Initial Setting of Bookinfo Application
+To demonstrate the scenario of using an external database, I start with a Kubernetes cluster with [Istio installed]({{home}}/docs/setup/kubernetes/quick-start.html#installation-steps). Then I deploy [Istio Bookinfo Sample Application]({{home}}/docs/guides/bookinfo.html). This application uses the _ratings_ microservice to fetch book ratings, a number between 1 and 5. The ratings are displayed as  ...
+
+The example commands in this blog post work with Istio version 0.3+, with or without [Mutual TLS](https://istio.io/docs/concepts/security/mutual-tls.html) enabled.
```
https://istio.io -> {{home}}

Here and elsewhere

geeknoid commented on _blog/2018/egress-tcp.md:
````diff
@@ -0,0 +1,306 @@
+### Use the Database for Ratings Data in Bookinfo Application
+1. I modify the deployment spec of a version of the _ratings_ microservice that uses a MySQL database, to use my database instance. The spec is in `samples/bookinfo/kube/bookinfo-ratings-v2-mysql.yaml` of an Istio release archive. I edit the following lines:
+
+    ```yaml
````
I think this is  indented one too many spaces. Should be like the other code blocks above. 

See the following few code blocks below, same issue.

frankbu commented on _blog/2018/egress-tcp.md:
```diff
@@ -0,0 +1,306 @@
+{% include home.html %}
+
+In my previous blog post, [Consuming External Web Services]({{home}}/blog/2018/egress-https.html), I described how external services can be consumed by in-mesh Istio applications via HTTPS. In this post, I demonstrate consuming external services over TCP. I use the [Istio Bookinfo sample application]({{home}}/docs/guides/bookinfo.html), the version in which the book ratings data is persisted in a ...
+
```
to allow the in-mesh application to access

frankbu commented on _blog/2018/egress-tcp.md:
```diff
@@ -0,0 +1,306 @@
+
+The example commands in this blog post work with Istio version 0.3+, with or without [Mutual TLS]({{home}}/docs/concepts/security/mutual-tls.html) enabled.
+
+Let me remind you the end-to-end architecture of the application from the original [Bookinfo Guide]({{home}}/docs/guides/bookinfo.html).
```
Let me remind you of the end-to-end ...

or maybe better:

As a reminder, here is the end-to-end architecture of the application from the [Bookinfo Guide]({{home}}/docs/guides/bookinfo.html).

frankbu commented on _blog/2018/egress-tcp.md:
````diff
@@ -0,0 +1,306 @@
+   Created config route-rule/default/reviews-test-ratings-v2 at revision 1918800
+   ```
+
+The updated architecture appears below. Note that the blue arrows inside the mesh mark the traffic allowed by the route rules we added. According to the route rules, the traffic is allowed to _reviews v3_ and _ratings v2-mysql_.
````
blue arrows inside the mesh mark the traffic enabled by the ...

also:

the traffic is sent to _reviews v3_ and _ratings v2-mysql_.

vadimeisenbergibm replied:
@frankbu Please note that the route rules do not enable the traffic, they limit the traffic. For example, without a rule to limit the traffic to _reviews v3_, the traffic would arrive to all the versions of _reviews_: _v1_, _v2_, _v3_.

vadimeisenbergibm replied:
@frankbu I replaced "allowed" by "sent".

frankbu replied:
Good point, I guess "configured" would have been a better choice, but "sent" is also OK.

frankbu commented on _blog/2018/egress-tcp.md:
```diff
@@ -0,0 +1,306 @@
+Next let's see how we define egress rules for TCP traffic.
+
+## Egress rules for TCP traffic
+The egress rules for enabling TCP traffic to a specific port must specify `TCP` as the protocol of the port. Additional non-HTTP TCP protocol currently supported is `MONGO`, the [MongoDB Wire Protocol](https://docs.mongodb.com/manual/reference/mongodb-wire-protocol/).
```
This is not clear ... does this mean that the egress rule must specify `TCP` for the port protocol OR, in the case of mongodb, it can (or must) set the port protocol to `MONGO`?

vadimeisenbergibm replied:
egress rules can specify either `TCP` for the port protocol, or (can, not must) specify `MONGO` in case of mongodb.

frankbu replied:
How about rephrasing like this:

Additionally, for the [MongoDB Wire Protocol](https://docs.mongodb.com/manual/reference/mongodb-wire-protocol/), the protocol can be specified as `MONGO`, instead of `TCP`.

vadimeisenbergibm replied:
Done, thanks!

ldemailly commented:
it would be useful I think to compare/contrast with the mesh expansion approach:
https://istio.io/docs/guides/integrating-vms.html

christian-posta commented:
before merging this, think we could get a squash of these commits @vadimeisenbergibm ?

vadimeisenbergibm commented:
@ldemailly Thanks, great idea! I have added a section "relation to mesh expansion". Could you please review it?

vadimeisenbergibm commented:
@christian-posta I think we always squash in Istio, in this case the squash should be done as well.

vadimeisenbergibm commented:
@christian-posta I meant we always squash **during the merge**.

ldemailly commented on _blog/2018/egress-tcp.md:
```diff
@@ -257,6 +257,11 @@ Note that all the IPs of an external service are not always known. To enable TCP
 Also note that the IPs of an external service are not always static, for example in the case of [CDNs](https://en.wikipedia.org/wiki/Content_delivery_network). Sometimes the IPs are static most of the time, but can be changed from time to time, for example due to infrastructure changes. In these cases, if the range of the possible IPs is known, you should specify the range by CIDR blocks, by mult ...
 
+## Relation to mesh expansion
+Note that the scenario described in this post is different from the mesh expansion scenario, described in the [Integrating Virtual Machines](https://istio.io/docs/guides/integrating-vms.html) guide. In that scenario, a MySQL instance runs on a an external (outside the cluster) machine (a bare metal or a VM), integrated with the Istio service mesh. The MySQL service becomes a first-class citizen o ...
```
Thanks for that. Can you add "including mTLS security" to the Among other things or benefits

vadimeisenbergibm replied:
@ldemailly Done, thanks!

vadimeisenbergibm commented:
@frankbu Regarding https://github.com/istio/istio.github.io/pull/944#discussion_r166773685,
> Good point, I guess "configured" would have been a better choice, but "sent" is also OK.

I have corrected this part to be:
Note that the blue arrows inside the mesh mark the traffic configured according to the route rules we added. According to the route rules, the traffic is sent to _reviews v3_ and _ratings v2-mysql_.

ldemailly commented:
@christian-posta and we also don't squash PRs after anyone has started looking (so people who did a first set of review can look only at deltas)
